plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java

```diff
@@ -169,7 +200,100 @@ static TrinoException handleException(ParquetDataSourceId dataSourceId, Exceptio
         return new TrinoException(HIVE_CURSOR_ERROR, format("Failed to read Parquet file: %s", dataSourceId), exception);
     }
 
-    private static Block getRowIndexColumn(long baseIndex, int size)
+    private static boolean isColumnAdaptationRequired(List<ColumnAdaptation> columnAdaptations)
+    {
+        // If no synthetic columns are added and the source columns are in order, no adaptations are required
+        for (int columnChannel = 0; columnChannel < columnAdaptations.size(); columnChannel++) {
+            ColumnAdaptation column = columnAdaptations.get(columnChannel);
+            if (column instanceof SourceColumn) {
+                int delegateChannel = ((SourceColumn) column).getSourceChannel();
+                if (columnChannel != delegateChannel) {
+                    return true;
+                }
+            }
+            else {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private interface ColumnAdaptation
+    {
+        Block getBlock(Page sourcePage, long startRowId);
+    }
+
+    private static class NullColumn
+            implements ColumnAdaptation
+    {
+        private final Block nullBlock;
+
+        private NullColumn(Type type)
+        {
+            this.nullBlock = type.createBlockBuilder(null, 1, 0)
+                    .appendNull()
+                    .build();
+        }
+
+        @Override
+        public Block getBlock(Page sourcePage, long startRowId)
+        {
+            return RunLengthEncodedBlock.create(nullBlock, sourcePage.getPositionCount());
+        }
+    }
+
+    private static class SourceColumn
+            implements ColumnAdaptation
+    {
+        private final int sourceChannel;
+
+        private SourceColumn(int sourceChannel)
+        {
+            checkArgument(sourceChannel >= 0, "sourceChannel is negative");
+            this.sourceChannel = sourceChannel;
+        }
+
+        @Override
+        public Block getBlock(Page sourcePage, long startRowId)
+        {
+            return sourcePage.getBlock(sourceChannel);
+        }
+
+        public int getSourceChannel()
+        {
+            return sourceChannel;
+        }
+    }
+
+    private static class ConstantColumn
+            implements ColumnAdaptation
+    {
+        private final Block singleValueBlock;
+
+        private ConstantColumn(Block singleValueBlock)
```

Comment:
`requireNonNull` on singleValueBlock?

Reply:
That's not required because we dereference it in the constructor, that will give null pointer exception with details automatically (nullptr exception messages are improved by default from idk 15 onwards).